### PR TITLE
nodenext compatibility

### PR DESCRIPTION
**Pull request by Uzlopak**
Well, the typings tests seem to run through... 

@fastify/typescript 

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

This scares me quite a lot

**Comment by Uzlopak** [comment]
Wy does it scare you?

**Comment by fox1t** [comment]
@mcollina I am terrified too. I rarely felt so scared by a PR.


@fastify/typescript what do you think about adding more tests for every possible target we support? The same as we have in the types playground?

**Comment by mcollina** [comment]
@kibertoad could you take a look as well?

**Comment by fox1t** [comment]
Before merging this, it would be nice to refactor the tsd test to support all of the possible combinations that we claim to support. It will also make us future-proof: we already saw how a slight change in TS can brake all our types. Another possible solution is to build a dedicated repo inside the org that consumes Fastify differently and outputs different targets.
 
@mcollina and @fastify/typescript, what do you think?

**Review by kibertoad — APPROVED** [review]

**Comment by Uzlopak** [comment]
I was thinking of a workflow

Something like (untested and not working):

```yaml
name: package-manager-ci

on:
  workflow_dispatch:
  push:
    branches:
      - main

permissions:
  contents: read

jobs:
  test:
    runs-on: ubuntu-latest
    permissions:
      contents: read

    strategy:
      matrix:
        module: [commonjs, node, nodenext]
        moduleResolution: [classic, node16, nodenext]
    steps:
      - uses: actions/checkout@v3
        with:
          persist-credentials: false

      - name: Use Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 14

      - name: Npm link current fastify-branch
        run: npm link

      - name: Clone Test Project
        run: git clone testproject

      - name: Test module ${{ matrix.module }} with ${{ matrix.moduleResolution }}
        run: |
          cd testproject &&
          npm i &&
          npm run typescript:${{ matrix.module }}:${{ matrix.moduleResolution }}

```

File: fastify.d.ts [hunk 1]
```diff
@@ -1,26 +1,190 @@
...
+  export interface ValidationResult {
+    keyword: string;
+    instancePath: string;
+    schemaPath: string;
+    params: Record<string, string | string[]>;
+    message?: string;
+  }
+
+  /* Export additional types */
+  export type {
```
Comment (climba03003) [inline fastify.d.ts:161]: I would clean up this export before landing.
In theory We can merge the export by using the same `namespace` in other file.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: fastify.d.ts (hunk 1 above)
Comment (Uzlopak) [reply fastify.d.ts:161]: You really want to use namespace for that?

**Comment by mcollina** [comment]
@climba03003 @Uzlopak any updates on this one?

File: fastify.d.ts (hunk 1 above)
Comment (climba03003) [reply fastify.d.ts:161]: It is better than managing massive of import and export.
Using `multi file namespace` allow to export in other file than merge all those in one.

Comment (Uzlopak) [reply fastify.d.ts:161]: It doesnt work as expected

Created a new PR with potential changes for this PR. 

https://github.com/fastify/fastify/pull/4491

Looking for feedback, on how we could potentially use namespace merging.

Comment (climba03003) [reply fastify.d.ts:161]: I will take a look tomorrow.

**Review by climba03003 — APPROVED** [review]
LGTM. 
I have tried once in refactoring for multi file namespace in #4491 .
Seems like a heavy job to let it works. So, it is better to merge this PR first and follow-up later.

**Pull request merged**